vm/systemSmartContracts/esdt.go

```diff
@@ -267,33 +251,204 @@ func (e *esdt) issueToken(owner []byte, arguments [][]byte) error {
-	//TODO: implement me
-	return vmcommon.Ok
+func (e *esdt) upgradeProperties(token *ESDTData, args [][]byte) {
+	for _, arg := range args {
+		optionalArg := string(arg)
+		switch optionalArg {
+		case burnable:
+			token.Burnable = true
+		case mintable:
+			token.Mintable = true
+		case canPause:
+			token.CanPause = true
+		case canFreeze:
+			token.CanFreeze = true
+		case canWipe:
+			token.CanWipe = true
+		case upgradable:
+			token.Upgradable = true
+		case canChangeOwner:
+			token.CanChangeOwner = true
+		}
+	}
 }
 
-func (e *esdt) mint(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	//TODO: implement me
+func (e *esdt) burn(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) != 2 {
+		e.eei.AddReturnMessage("number of arguments must be equal with 2")
+		return vmcommon.FunctionWrongSignature
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		e.eei.AddReturnMessage("callValue must be 0")
+		return vmcommon.OutOfFunds
+	}
+	err := e.eei.UseGas(e.gasCost.MetaChainSystemSCsCost.ESDTOperations)
+	if err != nil {
+		e.eei.AddReturnMessage("not enough gas")
+		return vmcommon.OutOfGas
+	}
+	burntValue := big.NewInt(0).SetBytes(args.Arguments[1])
+	if burntValue.Cmp(big.NewInt(0)) <= 0 {
+		e.eei.AddReturnMessage(vm.ErrNegativeOrZeroInitialSupply.Error())
+		return vmcommon.UserError
+	}
+	token, err := e.getExistingToken(args.Arguments[0])
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if !token.Burnable {
+		e.eei.AddReturnMessage("token is not burnable")
+		return vmcommon.UserError
+	}
+	token.BurntValue.Add(token.BurntValue, burntValue)
+
+	err = e.saveToken(token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
 	return vmcommon.Ok
 }
 
-func (e *esdt) freeze(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	//TODO: implement me
+func (e *esdt) mint(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) < 2 || len(args.Arguments) > 3 {
+		e.eei.AddReturnMessage("accepted arguments number 2/3")
+		return vmcommon.FunctionWrongSignature
+	}
+	token, returnCode := e.basicOwnerChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	mintValue := big.NewInt(0).SetBytes(args.Arguments[1])
+	if mintValue.Cmp(big.NewInt(0)) <= 0 {
+		e.eei.AddReturnMessage(vm.ErrNegativeOrZeroInitialSupply.Error())
+		return vmcommon.UserError
+	}
+	if !token.Mintable {
+		e.eei.AddReturnMessage("token is not burnable")
+		return vmcommon.UserError
+	}
+
+	token.MintedValue.Add(token.MintedValue, mintValue)
+	err := e.saveToken(token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	destination := token.OwnerAddress
+	if len(args.Arguments) == 3 {
+		if len(args.Arguments[2]) != len(args.CallerAddr) {
+			e.eei.AddReturnMessage("destination address of invalid length")
+			return vmcommon.UserError
+		}
+		destination = args.Arguments[2]
```

Comment:
why do we need the argument for the destination? Why not mint function should always sent tokens on the args.CallerAddr directly since it is the owner of the esdt token?

Reply:
You might want to send to a new contract or to a new investor.